Replace vagrant with docker

Earlopain opened the pull request:
First iteration of moving from vagrant to docker

Some improvements:
* You don't have to compile ruby. That removes a bunch of complexity because `ruby-install` and `chruby` are not needed. It also reduces first setup time since it doesn't need to be compiled. The base image is debian instead of ubuntu, though the process is basically the same.
* Storage persists through container reinstalls/updates because of volumes and can be destroyed individually.
* Live log output of all services when using docker-compose up.
* Seeding is no longer limited by the 30 per hour upoad limit. I needed the environment config to understand booleans for that.
* Storage usage should be improved.
* Hot reloading of `.erb` and `.rb` actually works now. Apparently this should already have been a thing but it didn't work with vagrant for some reason.
* Probably just a me problem: No more nfs fuckery. I couldn't execute `yarn install` inside the vm, it just got stuck.

Things to note:
* No hostmanager anymore. The instance must accessed with `localhost:3000`. I got it working by manualy editing /etc/hosts with static ips through compose on linux but couldn't get it to work on windows. I don't think it's even possible but I don't know that for certain.
* The postgres image creates a database named `danbooru`, just like the user. The actual database used is danbooru2`. If that database exists during `db:prepare` migrations are run which break horibly. I could not find a way to prevent postgres from creating the db when the volume is empty.
Windows requires some extra setup to get it working with reasonable performance. I added instructions for that to the readme. Developing on linux works fine without any additional steps.
* I removed a bunch of files/scripts which I believe were specific to danbooru.


Overall this seems to work quite nice. I gave it to @bitWolfy in advance to try and test it out and it just worked™. This is my first experience with docker and as such there are probably a few things that can be improved somehow.
@liamwhite I think you mentioned moving this to docker a few times in the past, so I assume you already have some experience with it. I would really appreciate if you could give me some input on this.

Earlopain committed:
[Misc] Remove unused install/docker files

Earlopain committed:
[Setup] Replace vagrant with docker

Earlopain committed:
[Tests] Remove unused images

Earlopain committed:
[Setup] Improve docker setup

* Reduces the image size by using alpine and cleaning up after some things
* Simplify the initial setup instructions slightly
* Access via localhost:3000 instead of e621.local

Earlopain committed:
[Setup] Ignore limits when seeding

This makes the environment config understand booleans. If something
ends with "?" it tries to interpret it as a boolean. A bunch of
existing configs where changed to reflect this change.

Earlopain committed:
[Setup] Save redis data to named volume

Prevents it from creating anonymous volumes over and over again

Earlopain committed:
[Readme] Add docker for windows specifics

liamwhite commented on Dockerfile:
```diff
@@ -0,0 +1,68 @@
+FROM ruby:2.7.3
```
See if you can get away with using alpine as a base instead. This will avoid you having to install a bunch of sources.list entries since alpine should already have everything you need, and the install size will be significantly smaller as well.

liamwhite commented on Dockerfile:
```diff
@@ -0,0 +1,68 @@
+  && tar xf vips-$VIPS_VERSION.tar.gz \
+  && cd vips-$VIPS_VERSION \
+  && ./configure --prefix=/usr \
+  && make install \
```
`make -j$(nproc) install`

liamwhite commented on bin/setup:
```diff
@@ -14,17 +14,20 @@ FileUtils.chdir APP_ROOT do
-  system! 'gem install bundler --conservative'
+  system! 'gem install bundler:2.0.1'
+  system! 'bundler config github.https true'
   system('bundle check') || system!('bundle install')
```
Should probably use `bundle install -j$(nproc)` as above to take advantage of parallelism.

liamwhite reviewed (changes requested):
1. Minor hazard: it did not work for me out of the box. I needed to run `sudo chcon -Rt svirt_sandbox_file_t .`, as the directory I used protects against bindmounts by default (SELinux). This is not the fault of the repository but you might want to be aware of this for troubleshooting purposes.

2. Make sure you add a `.dockerignore` file, otherwise you will get messages like this:
   ```
   Sending build context to Docker daemon  68.02MB
   ```
   where any local files you have (e.g. uploads) are influencing the build

Earlopain committed:
[Docker] Implement recommendations from review

* Use alpine as the base image
* Parallelize bundle install

Using alpine as base make is feasable to use the provides vips library.
The version is actually newer than what was installed previously.

Other stuff I did:
* Set SECRET_TOKEN/SESSION_SECRET_KEY from docker-compose
* Mute elasticsearch startup logspam
* Execute image as root. This means mounted volume do'nt need to  be chowned

Earlopain committed:
[Docker] Don't send build context

liamwhite commented on docker/postgres/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM postgres:12-alpine
+ARG BUILD_DEPS="git make gcc libc-dev clang llvm"
+
+RUN apk --no-cache add $BUILD_DEPS \
```
If you want to be able to rebuild this without purging your files whenever the test_parser repository changes, add this before the run command:

```
ADD https://api.github.com/repos/r888888888/test_parser/git/refs/heads/master /tmp/test_parser_version.json
```

liamwhite commented on docker/postgres/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+RUN apk --no-cache add $BUILD_DEPS \
+  && git clone https://github.com/r888888888/test_parser.git /tmp/test_parser \
+  && cd /tmp/test_parser \
+  && make install \
```
Parallel make? Not that it's strictly necessary for this one.

liamwhite reviewed (changes requested):
Much better, now it's significantly faster to build. Just a few small things:

Earlopain committed:
[Docker] A few small improvements

* Mark secrets as development only
* Parallelize test_parser build. I don't think there are enough files
to even hit the limit for me but good practice nontheless
* Rebuild postgres when test_parser changes. There has been no activity
in 6  years  and I don't expect this to change
